comps/llms/rayllm/api_openai_backend/router_app.py

```diff
@@ -0,0 +1,360 @@
+                        model=body.model,
+                        choices=choices,
+                        usage=None,
+                    )
+                    # data = chunk.json(exclude_unset=True, ensure_ascii=False)
+                    data = chunk.json()
+                    yield f"data: {data}\n\n"
+            if had_error:
+                # Return early in case of an error
+                break
+        if not had_error:
+            choices = [
+                DeltaChoices(
+                    delta=DeltaEOS(),
+                    index=0,
+                    finish_reason=finish_reason,
+                )
+            ]
+            usage = UsageInfo.from_response(ModelResponse.merge_stream(*all_results)) if all_results else None
+            chunk = ChatCompletionResponse(
+                id=completion_id,
+                object="chat.completion.result",
+                model=body.model,
+                choices=choices,
+                usage=usage,
+            )
+            data = chunk.json()
+            yield f"data: {data}\n\n"
+        yield "data: [DONE]\n\n"
+
+
+class Router:
+    def __init__(
+        self,
+        query_client: RouterQueryClient,
+    ) -> None:
+        self.query_client = query_client
+
+    @router_app.get("/v1/models", response_model=ModelList)
+    async def models(self) -> ModelList:
+        """OpenAI API-compliant endpoint to get all models."""
+        models = await self.query_client.models()
+        return ModelList(data=list(models.values()))
+
+    # :path allows us to have slashes in the model name
+    @router_app.get("/v1/models/{model:path}", response_model=ModelCard)
+    async def model_data(self, model: str) -> ModelCard:
+        """OpenAI API-compliant endpoint to get one model.
+
+        :param model: The model ID (e.g. "amazon/LightGPT")
+        """
+        model = model.replace("--", "/")
+        model_data = await self.query_client.model(model)
+        if model_data is None:
+            raise OpenAIHTTPException(
+                message=f"Invalid model '{model}'",
+                status_code=status.HTTP_400_BAD_REQUEST,
+                type="InvalidModel",
+            )
+        return model_data
+
+    @router_app.post("/v1/completions")
+    async def completions(
+        self,
+        body: CompletionRequest,
+        response: FastAPIResponse,
+    ):
+        """Given a prompt, the model will return one or more predicted completions,
+        and can also return the probabilities of alternative tokens at each position.
+
+        Returns:
+            A response object with completions.
+        """
+        prompt = Prompt(
+            prompt=body.prompt,
+            parameters=dict(body),
+            use_prompt_format=False,
+        )
+        request_id = f"cmpl-{str(uuid.uuid4().hex)}"
+
+        if body.stream:
+            return StreamingResponse(
+                _completions_wrapper(
```

Comment:
Completions API has been deprecated in favor of ChatCompletion API. Can consider not supporting it.